### PR TITLE
feat(event_sources): add CloudWatch dashboard custom widget event

### PR DESCRIPTION
**Pull request by sthuber90**
**Issue number:** #1461 

### Changes

> Please provide a summary of what's being changed

Added event mapping class for custom widget in CloudWatch dashboard

### User experience

> Please share what the user experience looks like before and after this change

With the new `CloudWatchDashboardCustomWidgetEvent` data class users have typed access to the [event](
https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/add_custom_widget_dashboard_about.html). The event will have custom parameters provided by the user in `event.params["name"]`, `event.raw_event["name"]` and `event["name"]`. If the Lambda function is invoked with the [`describe`](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/add_custom_widget_dashboard_about.html#add_custom_widget_describe) parameter, the function should return its documentation. The `describe` parameter can be provided alone without the `widgetContext` being part of the event or as a custom param by the invoking user.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [ ] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

-----
[View rendered docs/utilities/data_classes.md](https://github.com/sthuber90/aws-lambda-powertools-python/blob/feat-cw-dashboard-custom-widget-event-1461/docs/utilities/data_classes.md)

### CODE REVIEW

File: docs/utilities/data_classes.md [hunk 1]
```diff
@@ -441,6 +442,41 @@ In this example, we also use the new Logger `correlation_id` and built-in `corre
...
+    @event_source(data_class=CloudWatchDashboardCustomWidgetEvent)
+    def lambda_handler(event: CloudWatchDashboardCustomWidgetEvent, context):
+
+        if event.describe:
+            return DOCS
+
+        # alternatively you can also do
+        # event.widget_context.params["echo"]
+        # event.raw_event["echo"]
+        return event["echo"]
```
Comment (rubenfonseca) [inline docs/utilities/data_classes.md]: I haven't used this feature yet, but from the docs it seems that we could improve the docs by talking about the different return options:

```suggestion
        # You can directly return HTML or JSON content
        # Alternatively, you can return markdown that will be rendered by CloudWatch
        echo = event.widget_context.params["echo"]
        return { "markdown": f"# {echo}" }
```

Does it make sense?

**Review by rubenfonseca — DISMISSED** [review]
Thank you so much! Looks pretty good! Just left a small comment to improve the docs, let me know what you think :)

**Comment by leandrodamascena** [comment]
Hello @sthuber90! This is an awesome PR, thanks a lot for taking the time to help improve the project!

I think it's missing a field `domain` in event source schema.. Using [aws examples](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/add_custom_widget_samples.html) we can see the field `domain` in event source.

Could you please add this field?

image

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Add the `domain` field.

**Comment by sthuber90** [comment]
@leandrodamascena I looked at the link you posted and in the [cloudwatch-custom-widgets-samples](https://github.com/aws-samples/cloudwatch-custom-widgets-samples) repo but didn't find any `domain` (https://github.com/aws-samples/cloudwatch-custom-widgets-samples/search?q=domain). Furthermore, it's not mentioned in the [official docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/add_custom_widget_dashboard_about.html)

> Every call by a custom widget to Lambda includes a `widgetContext` element with the following contents, to provide the Lambda function developer with useful context information.

```
{
    "widgetContext": {
        "dashboardName": "Name-of-current-dashboard",
        "widgetId": "widget-16",
        "accountId": "012345678901",
        "locale": "en",
        "timezone": {
            "label": "UTC",
            "offsetISO": "+00:00",
            "offsetInMinutes": 0
        },
        "period": 300,
        "isAutoPeriod": true,
        "timeRange": {
            "mode": "relative",
            "start": 1627236199729,
            "end": 1627322599729,
            "relativeStart": 86400012,
            "zoom": {
                "start": 1627276030434,
                "end": 1627282956521
             }
        },
        "theme": "light",
        "linkCharts": true,
        "title": "Tweets for Amazon website problem",
        "forms": {
            "all": {}
        },
        "params": {
            "original": "param-to-widget"
        },
        "width": 588,
        "height": 369
    }
}
```

Can you please provide a link where the `domain` field is shown? And if we really want to add it, I think the AWS docs need to be updated as well.

**Comment by leandrodamascena** [comment]
Hi @sthuber90 - Answering your comments and questions:

> Can you please provide a link where the domain is shown?

There isn't. I've notified the documentation team to include it.

>  I looked at the link you posted and in the cloudwatch-custom-widgets-samples repo but didn't find any domain

I should've been more explicit in what examples contained the `domain` field - I've found them after deploying the "Echo", "Hello world" and "Simple SVG pie chart" examples.

As we've encountered this issue before (missing schema and docs accuracy), I've deployed all of those examples to confirm whether `domain` field was present - it was.

---

Steps to reproduce:

1 - I chose the "Echo",  "Hello World" and "SVG Pie Chart Example" examples  [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/add_custom_widget_samples.html).

2 - As all the examples are cloudformation templates, I just clicked on the link and deployed it to my account.

3 - When the stack was complete I went to Lambda Function and added a print to the event variable inside the handler function.
image

4 - I used the dashboard created earlier just to generate logs a few times. After that I could see in the Cloudwatch Lambda Log Groups the `domain` field in the event.
image

**Comment by sthuber90** [comment]
Thank you for the clarification. I will add the field in the next days 👍

**Comment by sthuber90** [comment]
@leandrodamascena added `domain` field

**Comment by leandrodamascena** [comment]
> @leandrodamascena added `domain` field

Hi @sthuber90!. I only removed a line in the example code and now it's ok to merge.

Thank you again.

**Review by leandrodamascena — APPROVED** [review]

**Pull request merged**